cluster-autoscaler/processors/nodeinfos/podtemplates/node_info_with_podtemplates_processors.go

```diff
@@ -0,0 +1,159 @@
+
+// NewNodeInfoWithPodTemplateProcessor returns a default instance of NodeInfoProcessor.
+func NewNodeInfoWithPodTemplateProcessor(opts *core.AutoscalerOptions) nodeinfos.NodeInfoProcessor {
+	internalContext, cancelFunc := context.WithCancel(context.Background())
+
+	return &nodeInfoWithPodTemplateProcessor{
+		ctx:               internalContext,
+		cancelFunc:        cancelFunc,
+		podTemplateLister: newPodTemplateLister(opts.KubeClient, internalContext.Done()),
+	}
+}
+
+// nodeInfoWithPodTemplateProcessor add possible PodTemplates in nodeInfos.
+type nodeInfoWithPodTemplateProcessor struct {
+	podTemplateLister v1lister.PodTemplateLister
+
+	ctx        context.Context
+	cancelFunc func()
+}
+
+// Process returns unchanged nodeInfos.
+func (p *nodeInfoWithPodTemplateProcessor) Process(ctx *ca_context.AutoscalingContext, nodeInfosForNodeGroups map[string]*schedulerframework.NodeInfo) (map[string]*schedulerframework.NodeInfo, error) {
+	// here we can use empty snapshot
+	clusterSnapshot := simulator.NewBasicClusterSnapshot()
+
+	// retrieve only once the podTemplates list.
+	// This list will be used for each NodeGroup.
+	podTemplates, err := p.podTemplateLister.List(labels.Everything())
+	if err != nil {
+		return nil, errors.ToAutoscalerError(errors.InternalError, err)
+	}
+
+	result := make(map[string]*schedulerframework.NodeInfo, len(nodeInfosForNodeGroups))
+	var errs []error
+	for id, nodeInfo := range nodeInfosForNodeGroups {
+		newNodeInfo, err := getNodeInfoWithPodTemplates(nodeInfo, podTemplates, clusterSnapshot, ctx.PredicateChecker)
+		if err != nil {
+			errs = append(errs, err)
+		}
+		result[id] = newNodeInfo
+	}
+
+	return result, utilerrors.NewAggregate(errs)
+}
+
+// CleanUp cleans up processor's internal structuxres.
+func (p *nodeInfoWithPodTemplateProcessor) CleanUp() {
+	p.cancelFunc()
+}
+
+func newPodTemplateLister(kubeClient client.Interface, stopchannel <-chan struct{}) v1lister.PodTemplateLister {
+	podTemplateWatchOption := func(options *metav1.ListOptions) {
+		options.FieldSelector = fields.Everything().String()
+		options.LabelSelector = labels.SelectorFromSet(getDaemonsetPodTemplateLabelSet()).String()
+	}
+	listWatcher := cache.NewFilteredListWatchFromClient(kubeClient.CoreV1().RESTClient(), "podtemplates", apiv1.NamespaceAll, podTemplateWatchOption)
+	store, reflector := cache.NewNamespaceKeyedIndexerAndReflector(listWatcher, &apiv1.PodTemplate{}, time.Hour)
+	lister := v1lister.NewPodTemplateLister(store)
+	go reflector.Run(stopchannel)
+	return lister
+}
+
+func getNodeInfoWithPodTemplates(baseNodeInfo *schedulerframework.NodeInfo, podTemplates []*apiv1.PodTemplate, clusterSnapshot *simulator.BasicClusterSnapshot, predicateChecker simulator.PredicateChecker) (*schedulerframework.NodeInfo, error) {
+	node := baseNodeInfo.Node()
+	var pods []*apiv1.Pod
+
+	for _, podInfo := range baseNodeInfo.Pods {
+		pods = append(pods, podInfo.Pod)
+	}
+	if err := clusterSnapshot.AddNodeWithPods(node, pods); err != nil {
```

Comment:
this seems like a side effect of calling this function, i would expect from the name that it just returns the nodeinfo with the pod templates. why do we add the node/pods to the cluster snapshot here?

Reply:
Thanks @elmiko  for the review.

I used the same logic that is present in the Daemonset support [here](https://github.com/kubernetes/autoscaler/blob/2542e8c884a8e25634d3b8f43243fa8706007f30/cluster-autoscaler/utils/daemonset/daemonset.go#L43) 
I need to check more deeply the purpose of this function.

Reply:
that makes sense following the daemonset pattern. i was mainly asking because i don't this code that well and was curious, i appreciate any further findings you might share =)

Reply:
Hello @elmiko,

Sorry for the late reply,

I checked why it is used like this in the `GetDaemonSetPodsForNode()` function. 

So for what I understand, it is because in this specific case the `Node` doesn't exist yet (we are in the edgecase where we scale from 0), and we don't want to add this Node the `context.AutoscalingContext.ClusterSnapshot` yet. 

The other solution could have been to use "context.AutoscalingContext.ClusterSnapshot.Fork()" but since we are not interested by the others `NodeInfo` here, creating a new `ClusterSnapshot` with `simulator.NewBasicClusterSnapshot()` limit the memory copy.

Reply:
thanks @clamoriniere ! i appreciate the extra context =)

Reply:
Using an empty snapshot implicitly makes nodes that are already in the clusters invisible to scheduling, in other words you implicitly assume such pods wouldn't impact the result of scheduling. There are some scheduling constraints (podAffinity, podTopologySpreading) that look at multiple nodes. Admittedly using any of those on DS or DS-like pod seems like a really bad idea, but technically I think it's more correct to go with Fork/Revert route. 

Also - Fork/Revert is used across codebase, but I don't think we actually use BasicClusterSnapshot anywhere in CA. So you're exposing yourself to code that is not tested in production.

Reply:
Hi @MaciekPytel 
thanks for the details. 
The edge-case that I would like to solve is the "scale from 0", for which the cluster-autoscaler is looking at Daemonset to simulate an existing Node. Thanks is why in this case, knowing the other existing Nodes doesn't seems useful. It is what I understood from code use to generate Pod based on Daemonset https://github.com/kubernetes/autoscaler/blob/2542e8c884a8e25634d3b8f43243fa8706007f30/cluster-autoscaler/utils/daemonset/daemonset.go#L43 which also use an empty snapshot. I thought it was an optimisation. 

But I don't see any reason why I can't change PodTemplate NodeInfoProcessor to use a Fork.